packages/gatsby-plugin-sitemap/src/internals.js

```diff
@@ -1,123 +1,222 @@
+      `You've passed something other than string to the exclude array. This is supported, but you'll have to write a custom filter function. 
+      Ignoring the input for now: ${JSON.stringify(excludedRoute, null, 2)}
+      https://www.gatsbyjs.com/plugins/gatsby-plugin-sitemap/#api-reference
+      `
     )
   }
 
-  // remove trailing slash of siteUrl
-  siteUrl = withoutTrailingSlash(siteUrl)
+  return minimatch(
+    withoutTrailingSlash(resolvePagePath(page)),
+    withoutTrailingSlash(excludedRoute)
+  )
+}
 
+/**
+ * @name serialize
+ *
+ * This function is executed by allPages.map(page => thisFunc(page, siteUrl, tools))
+ * allpages is the result of the filter process
+ *
+ * @param {object[]} page - results of the resolvePages function
+ * @param {object} tools - contains tools for serializing
+ *
+ */
+export function serialize(page, { resolvePagePath }) {
   return {
-    ...otherData,
-    allSitePage: {
-      [originalType]:
-        originalType === `nodes`
-          ? allPages
-          : allPages.map(page => {
-              return { node: page }
-            }),
-    },
-    site: { siteMetadata: { siteUrl } },
+    url: `${resolvePagePath(page)}`,
+    changefreq: `daily`,
+    priority: 0.7,
   }
 }
 
-export const defaultOptions = {
-  query: `
-    {
-      site {
-        siteMetadata {
-          siteUrl
-        }
-      }
+const defaultExcludes = [
+  `/dev-404-page`,
+  `/404`,
+  `/404.html`,
+  `/offline-plugin-app-shell-fallback`,
+]
+
+export function pageFilter({ allPages, filterPages, excludes }) {
+  const messages = []
+
+  if (
+    !Array.isArray(allPages) ||
+    typeof filterPages !== `function` ||
+    !Array.isArray(excludes)
+  ) {
+    throw new Error(`Invalid options passed to page Filter function`)
+  }
 
-      allSitePage {
-        edges {
-          node {
-            path
-          }
-        }
-      }
-  }`,
-  output: `/sitemap.xml`,
-  exclude: [
-    `/dev-404-page`,
-    `/404`,
-    `/404.html`,
-    `/offline-plugin-app-shell-fallback`,
-  ],
-  createLinkInHead: true,
-  serialize: ({ site, allSitePage }) => {
-    const { allPages } = getNodes(allSitePage)
-    return allPages?.map(page => {
-      return {
-        url: `${site.siteMetadata?.siteUrl ?? ``}${page.path}`,
-        changefreq: `daily`,
-        priority: 0.7,
+  const filteredPages = allPages.filter(page => {
+    // eslint-disable-next-line consistent-return
+    const defaultFilterMatches = defaultExcludes.some((exclude, i, arr) => {
+      try {
+        const doesMatch = defaultFilterPages(page, exclude, {
+          minimatch,
+          withoutTrailingSlash,
+          resolvePagePath,
+        })
+
+        //default excludes can only be found once, so remove them from the arr once excluded
+        if (doesMatch) arr.splice(i, 1)
+
+        return doesMatch
+      } catch {
+        throw new Error(`${ReporterPrefix} Error in default page filter`)
       }
     })
```

Comment:
With an explicit return we can remove the eslint disable comment
```suggestion
      }
      
      return false
    })
```